analytic_engine/src/instance/wal_replayer.rs

```diff
@@ -0,0 +1,596 @@
+
+        // Init all table results to be oks, and modify to errs when failed to replay.
+        let mut results = table_datas
+            .iter()
+            .map(|table_data| (table_data.id, Ok(())))
+            .collect();
+        let scan_ctx = ScanContext {
+            batch_size: context.wal_replay_batch_size,
+            ..Default::default()
+        };
+
+        Self::replay_region_logs(context, table_datas, &scan_ctx, &mut results).await?;
+
+        Ok(results)
+    }
+}
+
+impl RegionBasedCore {
+    /// Replay logs in same region.
+    ///
+    /// Steps:
+    /// + Scan all logs of region.
+    /// + Split logs according to table ids.
+    /// + Replay logs to recover data of tables.
+    async fn replay_region_logs(
+        context: &ReplayContext,
+        table_datas: &[TableDataRef],
+        scan_ctx: &ScanContext,
+        table_results: &mut HashMap<TableId, Result<()>>,
+    ) -> Result<()> {
+        // Scan all wal logs of current shard.
+        let scan_req = ScanRequest {
+            region_id: context.shard_id as RegionId,
+        };
+
+        let mut log_iter = context
+            .wal_manager
+            .scan(scan_ctx, &scan_req)
+            .await
+            .box_err()
+            .context(ReplayWalWithCause { msg: None })?;
+        let mut log_entry_buf = VecDeque::with_capacity(context.wal_replay_batch_size);
+
+        // Lock all related tables.
+        let mut serial_exec_ctxs = HashMap::with_capacity(table_datas.len());
+        for table_data in table_datas {
+            let serial_exec = table_data.serial_exec.lock().await;
+            let serial_exec_ctx = SerialExecContext {
+                table_data: table_data.clone(),
+                serial_exec,
+            };
+            serial_exec_ctxs.insert(table_data.id, serial_exec_ctx);
+        }
+
+        // Split and replay logs.
+        loop {
+            let decoder = WalDecoder::default();
+            log_entry_buf = log_iter
+                .next_log_entries(decoder, log_entry_buf)
+                .await
+                .box_err()
+                .context(ReplayWalWithCause { msg: None })?;
+
+            if log_entry_buf.is_empty() {
+                break;
+            }
+
+            Self::replay_single_batch(
+                context,
+                &log_entry_buf,
+                &mut serial_exec_ctxs,
+                table_results,
+            )
+            .await?;
+        }
+
+        Ok(())
+    }
+
+    async fn replay_single_batch(
+        context: &ReplayContext,
+        log_batch: &VecDeque<LogEntry<ReadPayload>>,
+        serial_exec_ctxs: &mut HashMap<TableId, SerialExecContext<'_>>,
+        table_results: &mut HashMap<TableId, Result<()>>,
+    ) -> Result<()> {
+        let mut table_batches = Vec::with_capacity(serial_exec_ctxs.len());
+        Self::split_log_batch_by_table(log_batch, &mut table_batches);
```

Comment:
Why not just
```rs
let table_batches = Self::split_log_batch_by_table(log_batch);
```
`table_batches`  capacity is not `serial_exec_ctxs.len()`, they have no relation at all.

Reply:
It is for reusing the `table_batches`, we don't need to reallocate the `table_batches` again and again while calling `split_log_batch_by_table` repeatedly.